overlord/snapstate/check_snap_test.go

```diff
@@ -425,6 +426,242 @@ version: 2
+
+	st.Unlock()
+	err = snapstate.CheckSnap(st, "snap-path", "kernel", nil, nil, snapstate.Flags{}, deviceCtx)
+	st.Lock()
+	c.Check(err, ErrorMatches, `cannot replace signed kernel snap with an unasserted one`)
+}
+
+func (s *checkSnapSuite) TestCheckUnassertedGadgetKernelSnapDangerousModel(c *C) {
+	reset := release.MockOnClassic(false)
+	defer reset()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	deviceCtx := &snapstatetest.TrivialDeviceContext{
+		DeviceModel: MakeModel20("gadget", map[string]interface{}{
+			"base":  "core20",
+			"grade": "dangerous",
+		}),
+	}
+	c.Check(deviceCtx.DeviceModel.Grade(), Equals, asserts.ModelDangerous)
+
+	s.setupKernelGadgetSnaps(st)
+
+	gadgetRestore := s.mockGadgetSnap(c, "")
+	defer gadgetRestore()
+
+	st.Unlock()
+	err := snapstate.CheckSnap(st, "snap-path", "gadget", nil, nil, snapstate.Flags{}, deviceCtx)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	kernelRestore := s.mockKernelSnap(c, "")
+	defer kernelRestore()
+
+	st.Unlock()
+	err = snapstate.CheckSnap(st, "snap-path", "kernel", nil, nil, snapstate.Flags{}, deviceCtx)
+	st.Lock()
+	c.Assert(err, IsNil)
+}
+
+func (s *checkSnapSuite) TestCheckAssertedGadgetKernelSnapSignedModel(c *C) {
+	reset := release.MockOnClassic(false)
+	defer reset()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	deviceCtx := &snapstatetest.TrivialDeviceContext{
+		DeviceModel: MakeModel20("gadget", map[string]interface{}{
+			"base":  "core20",
+			"grade": "signed",
+		}),
+	}
+	c.Check(deviceCtx.DeviceModel.Grade(), Equals, asserts.ModelSigned)
+
+	s.setupKernelGadgetSnaps(st)
+
+	gadgetRestore := s.mockGadgetSnap(c, "gadget-id")
+	defer gadgetRestore()
+
+	st.Unlock()
+	err := snapstate.CheckSnap(st, "snap-path", "gadget", nil, nil, snapstate.Flags{}, deviceCtx)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	kernelRestore := s.mockKernelSnap(c, "kernel-id")
+	defer kernelRestore()
+
+	st.Unlock()
+	err = snapstate.CheckSnap(st, "snap-path", "kernel", nil, nil, snapstate.Flags{}, deviceCtx)
+	st.Lock()
+	c.Assert(err, IsNil)
+}
+
+func (s *checkSnapSuite) TestCheckAssertedGadgetKernelSnapDangerousModel(c *C) {
+	reset := release.MockOnClassic(false)
+	defer reset()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	deviceCtx := &snapstatetest.TrivialDeviceContext{
+		DeviceModel: MakeModel20("gadget", map[string]interface{}{
+			"base":  "core20",
+			"grade": "dangerous",
+		}),
+	}
+	c.Check(deviceCtx.DeviceModel.Grade(), Equals, asserts.ModelDangerous)
+
+	s.setupKernelGadgetSnaps(st)
+
+	gadgetRestore := s.mockGadgetSnap(c, "gadget-id")
+	defer gadgetRestore()
+
+	st.Unlock()
+	err := snapstate.CheckSnap(st, "snap-path", "gadget", nil, nil, snapstate.Flags{}, deviceCtx)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	kernelRestore := s.mockKernelSnap(c, "kernel-id")
+	defer kernelRestore()
+
+	st.Unlock()
+	err = snapstate.CheckSnap(st, "snap-path", "kernel", nil, nil, snapstate.Flags{}, deviceCtx)
+	st.Lock()
+	c.Assert(err, IsNil)
+}
```

Comment:
Also all these tests here could be combined into a single data-driven test, since (unless I'm mistaken) the only changing parts are the input DeviceModel, the expected grade and the expected errors of CheckSnap().